mlir/lib/Quantum/Transforms/AdjointPatterns.cpp

```diff
@@ -0,0 +1,174 @@
+                        loc, insert.getQubit().getType(), query(insert.getOutQreg()),
+                        classicalMapping.lookupOrDefault(insert.getIdx()), insert.getIdxAttrAttr());
+                    update(insert.getQubit(), extract->getResult(0));
+                    update(insert.getInQreg(), quantumMapping[insert.getOutQreg()]);
+                }
+                else if (QuantumGate gate = dyn_cast<QuantumGate>(*i)) {
+                    IRMapping m(classicalMapping);
+                    for (const auto &[qr, qo] :
+                         llvm::zip(gate.getQubitResults(), gate.getQubitOperands())) {
+                        m.map(qo, query(qr));
+                    }
+                    QuantumGate clone = dyn_cast<QuantumGate>(rewriter.insert(i->clone(m)));
+                    clone.setAdjointFlag(!gate.getAdjointFlag());
+                    for (const auto &[qr, qo] :
+                         llvm::zip(clone.getQubitResults(), gate.getQubitOperands())) {
+                        update(qo, qr);
+                    }
+                }
+                else if (ExtractOp extract = dyn_cast<ExtractOp>(*i)) {
+                    auto insert = rewriter.create<InsertOp>(
+                        loc, extract.getQreg().getType(), query(extract.getQreg()),
+                        classicalMapping.lookupOrDefault(extract.getIdx()),
+                        extract.getIdxAttrAttr(), query(extract.getQubit()));
+                    update(extract.getQreg(), insert->getResult(0));
+                }
+                else if (AdjointOp adjoint2 = dyn_cast<AdjointOp>(*i)) {
+                    IRMapping m(classicalMapping);
+                    Block &b = adjoint2.getRegion().front();
+                    for (const auto &[a, r] : llvm::zip(b.getArguments(), adjoint2->getResults())) {
+                        m.map(a, query(r));
+                    }
+                    auto res = cloneAdjointRegion(adjoint2, rewriter, m);
+                    update(adjoint2.getQreg(), res);
+                }
+                else {
+                    /* TODO: We expect to handle Scf control flow instructions here. Stateless loops
+                     * and conditionals should not be a problem. Stateful loops would probably
+                     * require some kind of classical unrolling.
+                     */
+                }
+            }
+        };
+
+        // Finally, query and return the quantum outputs of the reversed program using the known
+        // input arguments of the source adjoint block as keys.
+        std::vector<Value> reversedOutputs;
+        {
+            for (auto a : adjoint.getRegion().front().getArguments()) {
+                reversedOutputs.push_back(quantumMapping[a]);
+            }
+        }
+
+        rewriter.replaceOp(adjoint, reversedOutputs);
+        return success();
+    }
+};
+
+} // namespace
+
+namespace catalyst {
+namespace quantum {
```

Comment:
nested namespaces can be merged since C++17. You can consider it here,
```suggestion
namespace catalyst::quantum {
```